src/Functions/array/arrayIntersect.cpp

```diff
@@ -549,44 +550,101 @@ ColumnPtr FunctionArrayIntersect::execute(const UnpackedArrays & arrays, Mutable
                 }
             }
 
-            prev_off[arg_num] = off;
-            if (arg.is_const)
-                prev_off[arg_num] = 0;
-
+            if (arg_num)
+            {
+                prev_off[arg_num] = off;
+                if (arg.is_const)
+                    prev_off[arg_num] = 0;
+            }
             if (!current_has_nullable)
                 all_has_nullable = false;
         }
 
-        if (all_has_nullable)
-        {
-            ++result_offset;
-            result_data.insertDefault();
-            null_map.push_back(1);
-        }
+        // We have NULL in output only once if it should be there
+        bool null_added = false;
+        const auto & arg = arrays.args[0];
+        size_t off;
+        // const array has only one row
+        if (arg.is_const)
+            off = (*arg.offsets)[0];
+        else
+            off = (*arg.offsets)[row];
 
-        for (const auto & pair : map)
+        for (auto i : collections::range(prev_off[0], off))
         {
-            if (pair.getMapped() == args)
+            all_has_nullable = all_nullable;
+            current_has_nullable = false;
+            typename Map::LookupResult pair = nullptr;
+
+            if (arg.null_map && (*arg.null_map)[i])
+            {
+                current_has_nullable = true;
+                if (null_added)
+                    continue;
+            }
+            else if constexpr (is_numeric_column)
+            {
+                pair = map.find(columns[0]->getElement(i));
+            }
+            else if constexpr (std::is_same_v<ColumnType, ColumnString> || std::is_same_v<ColumnType, ColumnFixedString>)
+                pair = map.find(columns[0]->getDataAt(i));
+            else
+            {
+                const char * data = nullptr;
+                pair = map.find(columns[0]->serializeValueIntoArena(i, arena, data));
+            }
+            prev_off[0] = off;
+            if (arg.is_const)
+                prev_off[0] = 0;
+
+            if (!current_has_nullable)
+                all_has_nullable = false;
+
+            if (pair && pair->getMapped() == args)
             {
+                ++pair->getMapped();
```

Comment:
Let's add a comment here as well.
We increase `pair->getMapped()` here to not skip duplicate values from the first array.